transport/x/grpc/peer.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"context"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/peer/hostport"
+	"go.uber.org/yarpc/yarpcerrors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+)
+
+type grpcPeer struct {
+	*hostport.Peer
+	t          *Transport
+	clientConn *grpc.ClientConn
+	stoppingC  chan struct{}
+	stoppedC   chan error
+}
+
+func newPeer(address string, t *Transport) (*grpcPeer, error) {
+	clientConn, err := grpc.Dial(
+		address,
+		grpc.WithInsecure(),
+		grpc.WithCodec(customCodec{}),
+		grpc.WithUserAgent(UserAgent),
+	)
+	if err != nil {
+		return nil, err
+	}
+	grpcPeer := &grpcPeer{
+		Peer:       hostport.NewPeer(hostport.PeerIdentifier(address), t),
+		t:          t,
+		clientConn: clientConn,
+		stoppingC:  make(chan struct{}, 1),
+		stoppedC:   make(chan error, 1),
+	}
+	go grpcPeer.monitor()
+	return grpcPeer, nil
+}
+
+func (p *grpcPeer) monitor() {
+	// wait for start so we can be certain that we have a channel
```

Comment:
This actually verifies I call transport.Start() in testing, so I'd want to keep it anyways :-)